drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -135,6 +122,80 @@ static void adc_disable_measurement(void)
 	irq_disable(DT_INST_IRQN(0));
 }
 
+/* Get result for each ADC selected channel. */
+static void adc_it8xxx2_get_sample(const struct device *dev)
+{
+	struct adc_it8xxx2_data *data = dev->data;
+	struct adc_it8xxx2_regs *const adc_regs = ADC_IT8XXX2_REG_BASE;
+
+	if (adc_regs->VCH0CTL & IT8XXX2_ADC_DATVAL) {
+		/* Read adc raw data of msb and lsb */
+		*data->buffer++ = adc_regs->VCH0DATM << 8 | adc_regs->VCH0DATL;
+	} else {
+		LOG_WRN("ADC failed to read (regs=%x, ch=%d)",
+			adc_regs->ADCDVSTS, data->ch);
+	}
+
+	adc_disable_measurement();
+}
+
+static void adc_enable_measurement(uint32_t ch)
+{
+	struct gctrl_it8xxx2_regs *const gctrl_base =
+		(struct gctrl_it8xxx2_regs *)
+			DT_REG_ADDR(DT_NODELABEL(gctrl));
+	struct adc_it8xxx2_regs *const adc_regs = ADC_IT8XXX2_REG_BASE;
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	struct adc_it8xxx2_data *data = dev->data;
+
+	/* Select and enable a voltage channel input for measurement */
+	adc_regs->VCH0CTL = (IT8XXX2_ADC_DATVAL | IT8XXX2_ADC_INTDVEN) + ch;
+
+	/* ADC module enable */
+	adc_regs->ADCCFG |= IT8XXX2_ADC_ADCEN;
+
+	/*
+	 * In the sampling process, it is possible to read multiple channels
+	 * at a time. The ADC sampling of it8xxx2 needs to read each channel
+	 * in sequence, so it needs to wait for an interrupt to read data in
+	 * the loop through k_sem_take(). But k_timer_start() is used in the
+	 * interval test in test_adc.c, so we need to use polling wait instead
+	 * of k_sem_take() to wait, otherwise it will cause kernel panic.
+	 *
+	 * k_is_in_isr() can determine whether to use polling or k_sem_take()
+	 * at present.
+	 */
+	if (k_is_in_isr()) {
+		int delay_count = 0;
+
+		while (!(adc_regs->VCH0CTL & IT8XXX2_ADC_DATVAL)) {
```

Comment:
we're getting pretty nested here. consider moving some of this to a helper function

Reply:
Done.